dev/tools/localization/gen_l10n.dart

```diff
@@ -295,67 +109,190 @@ String generatePluralMethod(Message message) {
+  if (message.placeholders.isNotEmpty) {
+    return baseClassMethodTemplate
+      .replaceAll('@(comment)', comment)
+      .replaceAll('@(name)', message.resourceId)
+      .replaceAll('@(parameters)', generateMethodParameters(message).join(', '));
+  }
+  return baseClassGetterTemplate
+    .replaceAll('@(comment)', comment)
+    .replaceAll('@(name)', message.resourceId);
+}
+
+class LocalesForLanguage {
+  LocalesForLanguage(this.languageCode);
+  final String languageCode;
+  final List<LocaleInfo> locales = <LocaleInfo>[];
+  final List<String> scriptCodes = <String>[];
+  final List<String> countryCodes = <String>[];
+
+  Iterable<LocaleInfo> localesWithCountryCode(String countryCode) {
+    return locales.where((LocaleInfo locale) => locale.countryCode == countryCode);
+  }
+
+  Iterable<LocaleInfo> localesWithScriptCode(String scriptCode) {
+    return locales.where((LocaleInfo locale) => locale.scriptCode == scriptCode);
+  }
+}
+
+String generateLookupBody(Iterable<LocaleInfo> allLocales, String className) {
+  final Map<String, LocalesForLanguage> languageToLocales = <String, LocalesForLanguage>{};
+  for (final LocaleInfo locale in allLocales) {
+    final String languageCode = locale.languageCode;
+    languageToLocales[languageCode] ??= LocalesForLanguage(languageCode);
+    languageToLocales[languageCode].locales.add(locale);
+    if (locale.scriptCode != null)
+      languageToLocales[languageCode].scriptCodes.add(locale.scriptCode);
+    if (locale.countryCode != null)
+      languageToLocales[languageCode].countryCodes.add(locale.countryCode);
+  }
+
+  final Iterable<String> switchClauses = languageToLocales.values.map((LocalesForLanguage localesForLanguage) {
+    if (localesForLanguage.scriptCodes.isEmpty && localesForLanguage.countryCodes.isEmpty) {
+      assert(localesForLanguage.locales.length == 1);
+      final LocaleInfo locale = localesForLanguage.locales[0];
+      return switchClauseTemplate
+        .replaceAll('@(case)', localesForLanguage.languageCode)
+        .replaceAll('@(class)', '$className${locale.camelCase()}');
+    }
+
+    if (localesForLanguage.scriptCodes.isEmpty) {
+      return localesForLanguage.countryCodes.map((String countryCode) {
+        final Iterable<LocaleInfo> locales = localesForLanguage.localesWithCountryCode(countryCode);
+        return countryCodeSwitchTemplate
+          .replaceAll('@(languageCode)', localesForLanguage.languageCode)
+          .replaceAll('@(class)', '$className${LocaleInfo.fromString(localesForLanguage.languageCode).camelCase()}')
+          .replaceAll('@(switchClauses)', locales.map((LocaleInfo locale) {
+            return switchClauseTemplate
+              .replaceAll('@(case)', countryCode)
+              .replaceAll('@(class)', '$className${locale.camelCase()}');
+          }).join('\n'));
+      }).join('\n');
+    }
+    assert(false, 'TBD dispatch on scriptCodes, scriptCodes AND countryCodes');
```

Comment:
Is it ok to leave this as an assert? It we do, maybe we should rephrase it to suggest what is possible.

Also, when should we expect the tool to handle script codes as well?

Reply:
It should handle script codes, although we don't have tests cases to cover that yet. I will add this stuff; didn't mean to leave it as a TBD.